machine-learning/ml/preprocessing/preprocessing.py

```diff
@@ -63,38 +48,41 @@ def retrieve_labelled_instances(dataset, refactoring: LowLevelRefactoring):
         non_refactored_instances = non_refactored_instances.sample(frac=0.1)
 
     # now, combine both datasets (with both TRUE and FALSE predictions)
-    assert non_refactored_instances.shape[1] == refactored_instances.shape[1], "number of columns differ from both datasets"
+    if non_refactored_instances.shape[1] != refactored_instances.shape[1]:
+        raise ImportError("Number of columns differ from both datasets.")
     merged_dataset = pd.concat([refactored_instances, non_refactored_instances])
 
     # just to be sure, shuffle the dataset
     merged_dataset = merged_dataset.sample(frac=1, random_state = 42)
 
+    # do we want to try the models without some metrics, e.g. process and authorship metrics?
+    merged_dataset = merged_dataset.drop(DROP_METRICS, axis=1)
+
     # separate the x from the y (as required by the scikit-learn API)
     x = merged_dataset.drop("prediction", axis=1)
     y = merged_dataset["prediction"]
 
-    # do we want to try the models without process and authorship metrics?
-    if DROP_PROCESS_AND_AUTHORSHIP_METRICS:
-        x = x.drop(["authorOwnership", "bugFixCount", "qtyMajorAuthors",
-                    "qtyMinorAuthors", "qtyOfAuthors", "qtyOfCommits", "refactoringsInvolved"], axis=1)
-
     # balance the datasets, as we have way more 'non refactored examples' rather than refactoring examples
     # for now, we basically perform under sampling
-    if BALANCE_DATASET:
-        log("instances before balancing: {}".format(Counter(y)))
+    if is_training_data and BALANCE_DATASET:
+        log("instances before balancing: {}".format(Counter(y)), False)
         x, y = perform_balancing(x, y)
         assert x.shape[0] == y.shape[0], "Balancing did not work, x and y have different shapes."
-        log("instances after balancing: {}".format(Counter(y)))
+        log("instances after balancing: {}".format(Counter(y)), False)
 
     # apply some scaling to speed up the algorithm
-    scaler = None
-    if SCALE_DATASET:
+    if SCALE_DATASET and scaler is None:
         x, scaler = perform_scaling(x)
+    elif SCALE_DATASET and scaler is not None:
```

Comment:
Do we need to replace the scaler here? I'm supposing it returns the same one, right?